cpp/src/arrow/util/parsing.h

```diff
@@ -108,81 +109,213 @@ class StringConverter<DoubleType> : public StringToFloatConverterMixin<DoubleTyp
 
 // NOTE: HalfFloatType would require a half<->float conversion library
 
+namespace detail {
+
+#define PARSE_UNSIGNED_ITERATION(C_TYPE)              \
+  if (length > 0) {                                   \
+    uint8_t digit = static_cast<uint8_t>(*s++ - '0'); \
+    result = static_cast<C_TYPE>(result * 10U);       \
+    length--;                                         \
+    if (ARROW_PREDICT_FALSE(digit > 9U)) {            \
+      /* Non-digit */                                 \
+      return false;                                   \
+    }                                                 \
+    result = static_cast<C_TYPE>(result + digit);     \
+  }
+
+#define PARSE_UNSIGNED_ITERATION_LAST(C_TYPE)                                     \
+  if (length > 0) {                                                               \
+    if (ARROW_PREDICT_FALSE(result > std::numeric_limits<C_TYPE>::max() / 10U)) { \
+      /* Overflow */                                                              \
```

Comment:
At some point do you think we might want to return a "reason" as an out argument when the parser returns false? Could be follow up work

Reply:
Not sure, because if we're calling out to an external parser (as we'll keep doing for floating-point parsing), it may not give us a reason.